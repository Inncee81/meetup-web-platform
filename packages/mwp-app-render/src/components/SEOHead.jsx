import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';

import { DEFAULT_TITLE } from 'mwp-app-render/src/util/seo/ldJson';
import { generateCanonicalUrlMetaTags } from 'mwp-app-render/src/util/seo/links';
import {
	DEFAULT_IMAGE_URL,
	generateMetaData,
	generateMetaTags,
} from 'mwp-app-render/src/util/seo/metaTags';

/**
 * Component for rendering SEO content in the document's head
 * Note: props passed to SEOHead should *not* be html-escaped because this is handled by Helmet
 * @see https://github.com/nfl/react-helmet
 * @param {String} baseUrl Base url of the page (protocol + hostname).
 * @param {String} imageUrl An image url that will be applied to the link tag, og:image meta tag, etc.
 * @param {Array} ldJson Array containing json-ld objects. These are usually generated by helper functions in src/util/seoHelper.js.
 * @param {String} localCode The current localeCode.
 * @param {String} ogDescription A special description for the Open Graph og:description meta tag. If this isn't provided pageDescription is used.
 * @param {String} ogTitle A special title for the Open Graph og:title meta tag. If this isn't provided pageTitle is used.
 * @param {String} pageDescription The page description, for example -- Upcoming Events for NY Tech in New York City
 * @param {String} pageKeywords A comma-separated string of keywords, for example -- 'upcoming events,ny-tech,ny,technology,group,club,event,community,local,networking,meet,sharing'
 * @param {Array} pageMeta Optional meta tag data, for example -- [{name: 'yandex', value: 12345},{name: 'bing', value: 12345}]
 * @param {String} pageTitle Content for the page's title tag
 * @param {Boolean} robots Instructs search engines whether or not we'd like them to crawl the pages and its links
 * @param {String} route The current route
 * @module SEOHead
 */
export const SEOHeadComponent = ({
	baseUrl,
	imageUrl,
	ldJson,
	localeCode,
	ogDescription,
	ogTitle,
	pageDescription,
	pageKeywords,
	pageMeta,
	pageTitle,
	robots,
	route,
}) => {

	const metaData = generateMetaData({
		appPath: `meetup:/${route}`,
		baseUrl,
		description: pageDescription,
		imageUrl,
		localeCode,
		keywords: pageKeywords,
		ogTitle,
		ogDescription,
		route,
		title: pageTitle,
	});

	const metaTags = generateMetaTags([...metaData, ...pageMeta]);

	const canonicalUrlMetaTags = generateCanonicalUrlMetaTags(baseUrl, localeCode, route);

	const ldJsonTags = ldJson.map((jsonObj, index) => (
		// eslint-disable-next-line react/no-array-index-key
		<script type="application/ld+json" key={`ldjson-${index}`}>
			{JSON.stringify(jsonObj)}
		</script>
	));

	return (
		<Helmet defaultTitle="Meetup" titleTemplate="%s | Meetup">
			<title>{pageTitle}</title>
			<link rel="image_src" href={imageUrl} />

			{robots ? (
				<meta name="robots" content="index, follow" />
			) : (
				<meta name="robots" content="noindex, nofollow" />
			)}

			{metaTags}

			{canonicalUrlMetaTags}

			{ldJsonTags}
		</Helmet>
	);
};

SEOHeadComponent.propTypes = {
	baseUrl: PropTypes.string.isRequired,
	imageUrl: PropTypes.string,
	ldJson: PropTypes.array,
	localeCode: PropTypes.string,
	ogTitle: PropTypes.string,
	ogDescription: PropTypes.string,
	pageDescription: PropTypes.string,
	pageKeywords: PropTypes.string,
	pageMeta: PropTypes.array,
	pageTitle: PropTypes.string,
	robots: PropTypes.bool,
	route: PropTypes.string.isRequired,
};

SEOHeadComponent.defaultProps = {
	pageMeta: [],
	pageTitle: DEFAULT_TITLE,
	imageUrl: DEFAULT_IMAGE_URL,
	ldJson: [],
	localeCode: 'en-US',
	robots: true,
};

export default SEOHeadComponent;
