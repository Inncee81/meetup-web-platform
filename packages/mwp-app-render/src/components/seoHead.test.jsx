import React from 'react';
import { shallow } from 'enzyme';
import { SEOHeadComponent } from './SEOHead';

const MOCK_PROPS = {
	baseUrl: 'https://baseUrl.com',
	imageUrl: 'https://www.mock-image-url.com/',
	ldJson: [{ foo: 'MOCK_LD_JSON' }, { bar: 'MOCK_LD_JSON' }],
	localeCode: 'es-ES',
	pageDescription: 'mock description',
	pageKeywords: 'mock keywords',
	pageMeta: [{ mock: 'meta' }, { more: 'mock' }],
	pageTitle: 'mock-page-title',
	robots: false,
	route: '/mock/preview/route',
};

const renderComponent = (props = MOCK_PROPS) => shallow(<SEOHeadComponent {...props} />);

describe('SEOHead', () => {
	it('renders component markup', () => {
		expect(renderComponent()).toMatchSnapshot();
	});
	describe('robots', () => {
		const robotsSelector = 'meta[name="robots"]';
		it('defaults to robots "on"', () => {
			const props = {...MOCK_PROPS};
			delete props.robots;
			expect(renderComponent(props).find(robotsSelector)).toMatchSnapshot();
		});
		it('explicity turns robots "on"', () => {
			const props = {...MOCK_PROPS, robots: true};
			delete props.robots;
			expect(renderComponent(props).find(robotsSelector)).toMatchSnapshot();
		});
		it('turns robots "off"', () => {
			const props = {...MOCK_PROPS, robots: false};
			expect(renderComponent(props).find(robotsSelector)).toMatchSnapshot();
		});
	});
	describe('Custom Open Graph data', () => {
		it('renders custom open graph description', () => {
			const props = {...MOCK_PROPS, ogDescription: 'mock custom open graph description' };
			expect(renderComponent(props).find('meta[property="og:description"]')).toMatchSnapshot();
		});
	});
	describe('imageUrl', () => {
		it('applies meet script logo to og:image, twitter:image and img_src tags when imageUrl is not provided', () => {
			const props = {
				...MOCK_PROPS,
				imageUrl: undefined,
			};
			expect(renderComponent(props)).toMatchSnapshot();
		});
	});
});
