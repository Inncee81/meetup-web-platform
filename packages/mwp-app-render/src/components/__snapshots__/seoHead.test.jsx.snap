// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEOHead Custom Open Graph data renders custom open graph description 1`] = `
<meta
  content="mock custom open graph description"
  key="meta-6"
  property="og:description"
/>
`;

exports[`SEOHead imageUrl applies meet script logo to og:image, twitter:image and img_src tags when imageUrl is not provided 1`] = `
<HelmetWrapper
  defaultTitle="Meetup"
  defer={true}
  encodeSpecialCharacters={true}
  titleTemplate="%s | Meetup"
>
  <title>
    mock-page-title
  </title>
  <link
    href="https://secure.meetupstatic.com/s/img/286374644891845767035/logo/meetup-logo-script-1200x630.png"
    rel="image_src"
  />
  <meta
    content="noindex, nofollow"
    name="robots"
  />
  <meta
    content="mock description"
    key="meta-0"
    name="description"
  />
  <meta
    content="mock keywords"
    key="meta-1"
    name="keywords"
  />
  <meta
    content="36212898404"
    key="meta-2"
    property="fb:app_id"
  />
  <meta
    content="Meetup"
    key="meta-3"
    property="og:site_name"
  />
  <meta
    content="article"
    key="meta-4"
    property="og:type"
  />
  <meta
    content="mock-page-title"
    key="meta-5"
    property="og:title"
  />
  <meta
    content="mock description"
    key="meta-6"
    property="og:description"
  />
  <meta
    content="https://baseUrl.com/es-ES/mock/preview/route"
    key="meta-7"
    property="og:url"
  />
  <meta
    content="https://secure.meetupstatic.com/s/img/286374644891845767035/logo/meetup-logo-script-1200x630.png"
    key="meta-8"
    property="og:image"
  />
  <meta
    content="Meetup"
    key="meta-9"
    property="al:android:app_name"
  />
  <meta
    content="com.meetup"
    key="meta-10"
    property="al:android:package"
  />
  <meta
    content="375990038"
    key="meta-11"
    property="al:ios:app_store_id"
  />
  <meta
    content="Meetup"
    key="meta-12"
    property="al:ios:app_name"
  />
  <meta
    content="true"
    key="meta-13"
    property="al:web:should_fallback"
  />
  <meta
    content="meetup://mock/preview/route"
    key="meta-14"
    property="al:android:url"
  />
  <meta
    content="meetup://mock/preview/route"
    key="meta-15"
    property="al:ios:url"
  />
  <meta
    content="https://baseUrl.com/es-ES/mock/preview/route"
    key="meta-16"
    property="al:web:url"
  />
  <meta
    content="summary_large_image"
    key="meta-17"
    property="twitter:card"
  />
  <meta
    content="@meetup"
    key="meta-18"
    property="twitter:site"
  />
  <meta
    content="https://secure.meetupstatic.com/s/img/286374644891845767035/logo/meetup-logo-script-1200x630.png"
    key="meta-19"
    property="twitter:image"
  />
  <meta
    content="mock description"
    key="meta-20"
    property="twitter:description"
  />
  <meta
    content="375990038"
    key="meta-21"
    property="twitter:app:id:iphone"
  />
  <meta
    content="Meetup"
    key="meta-22"
    property="twitter:app:name:iphone"
  />
  <meta
    content="375990038"
    key="meta-23"
    property="twitter:app:id:ipad"
  />
  <meta
    content="Meetup"
    key="meta-24"
    property="twitter:app:name:ipad"
  />
  <meta
    content="com.meetup"
    key="meta-25"
    property="twitter:app:id:googleplay"
  />
  <meta
    content="Meetup"
    key="meta-26"
    property="twitter:app:name:googleplay"
  />
  <meta
    key="meta-27"
    mock="meta"
  />
  <meta
    key="meta-28"
    more="mock"
  />
  <link
    href="https://baseUrl.com/mock/preview/route"
    hrefLang="en"
    key="en-US"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/de-DE/mock/preview/route"
    hrefLang="de-DE"
    key="de-DE"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/en-AU/mock/preview/route"
    hrefLang="en-AU"
    key="en-AU"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/es/mock/preview/route"
    hrefLang="es"
    key="es"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/es-ES/mock/preview/route"
    hrefLang="es-ES"
    key="es-ES"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/fr-FR/mock/preview/route"
    hrefLang="fr-FR"
    key="fr-FR"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/it-IT/mock/preview/route"
    hrefLang="it-IT"
    key="it-IT"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/ja-JP/mock/preview/route"
    hrefLang="ja-JP"
    key="ja-JP"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/ko-KR/mock/preview/route"
    hrefLang="ko-KR"
    key="ko-KR"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/nl-NL/mock/preview/route"
    hrefLang="nl-NL"
    key="nl-NL"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/pt-BR/mock/preview/route"
    hrefLang="pt-BR"
    key="pt-BR"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/pl-PL/mock/preview/route"
    hrefLang="pl-PL"
    key="pl-PL"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/ru-RU/mock/preview/route"
    hrefLang="ru-RU"
    key="ru-RU"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/th-TH/mock/preview/route"
    hrefLang="th-TH"
    key="th-TH"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/tr-TR/mock/preview/route"
    hrefLang="tr-TR"
    key="tr-TR"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/mock/preview/route"
    hrefLang="x-default"
    key="default"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/es-ES/mock/preview/route"
    key="canonical"
    rel="canonical"
  />
  <script
    key="ldjson-0"
    type="application/ld+json"
  >
    {"foo":"MOCK_LD_JSON"}
  </script>
  <script
    key="ldjson-1"
    type="application/ld+json"
  >
    {"bar":"MOCK_LD_JSON"}
  </script>
</HelmetWrapper>
`;

exports[`SEOHead renders component markup 1`] = `
<HelmetWrapper
  defaultTitle="Meetup"
  defer={true}
  encodeSpecialCharacters={true}
  titleTemplate="%s | Meetup"
>
  <title>
    mock-page-title
  </title>
  <link
    href="https://www.mock-image-url.com/"
    rel="image_src"
  />
  <meta
    content="noindex, nofollow"
    name="robots"
  />
  <meta
    content="mock description"
    key="meta-0"
    name="description"
  />
  <meta
    content="mock keywords"
    key="meta-1"
    name="keywords"
  />
  <meta
    content="36212898404"
    key="meta-2"
    property="fb:app_id"
  />
  <meta
    content="Meetup"
    key="meta-3"
    property="og:site_name"
  />
  <meta
    content="article"
    key="meta-4"
    property="og:type"
  />
  <meta
    content="mock-page-title"
    key="meta-5"
    property="og:title"
  />
  <meta
    content="mock description"
    key="meta-6"
    property="og:description"
  />
  <meta
    content="https://baseUrl.com/es-ES/mock/preview/route"
    key="meta-7"
    property="og:url"
  />
  <meta
    content="https://www.mock-image-url.com/"
    key="meta-8"
    property="og:image"
  />
  <meta
    content="Meetup"
    key="meta-9"
    property="al:android:app_name"
  />
  <meta
    content="com.meetup"
    key="meta-10"
    property="al:android:package"
  />
  <meta
    content="375990038"
    key="meta-11"
    property="al:ios:app_store_id"
  />
  <meta
    content="Meetup"
    key="meta-12"
    property="al:ios:app_name"
  />
  <meta
    content="true"
    key="meta-13"
    property="al:web:should_fallback"
  />
  <meta
    content="meetup://mock/preview/route"
    key="meta-14"
    property="al:android:url"
  />
  <meta
    content="meetup://mock/preview/route"
    key="meta-15"
    property="al:ios:url"
  />
  <meta
    content="https://baseUrl.com/es-ES/mock/preview/route"
    key="meta-16"
    property="al:web:url"
  />
  <meta
    content="summary_large_image"
    key="meta-17"
    property="twitter:card"
  />
  <meta
    content="@meetup"
    key="meta-18"
    property="twitter:site"
  />
  <meta
    content="https://www.mock-image-url.com/"
    key="meta-19"
    property="twitter:image"
  />
  <meta
    content="mock description"
    key="meta-20"
    property="twitter:description"
  />
  <meta
    content="375990038"
    key="meta-21"
    property="twitter:app:id:iphone"
  />
  <meta
    content="Meetup"
    key="meta-22"
    property="twitter:app:name:iphone"
  />
  <meta
    content="375990038"
    key="meta-23"
    property="twitter:app:id:ipad"
  />
  <meta
    content="Meetup"
    key="meta-24"
    property="twitter:app:name:ipad"
  />
  <meta
    content="com.meetup"
    key="meta-25"
    property="twitter:app:id:googleplay"
  />
  <meta
    content="Meetup"
    key="meta-26"
    property="twitter:app:name:googleplay"
  />
  <meta
    key="meta-27"
    mock="meta"
  />
  <meta
    key="meta-28"
    more="mock"
  />
  <link
    href="https://baseUrl.com/mock/preview/route"
    hrefLang="en"
    key="en-US"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/de-DE/mock/preview/route"
    hrefLang="de-DE"
    key="de-DE"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/en-AU/mock/preview/route"
    hrefLang="en-AU"
    key="en-AU"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/es/mock/preview/route"
    hrefLang="es"
    key="es"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/es-ES/mock/preview/route"
    hrefLang="es-ES"
    key="es-ES"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/fr-FR/mock/preview/route"
    hrefLang="fr-FR"
    key="fr-FR"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/it-IT/mock/preview/route"
    hrefLang="it-IT"
    key="it-IT"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/ja-JP/mock/preview/route"
    hrefLang="ja-JP"
    key="ja-JP"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/ko-KR/mock/preview/route"
    hrefLang="ko-KR"
    key="ko-KR"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/nl-NL/mock/preview/route"
    hrefLang="nl-NL"
    key="nl-NL"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/pt-BR/mock/preview/route"
    hrefLang="pt-BR"
    key="pt-BR"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/pl-PL/mock/preview/route"
    hrefLang="pl-PL"
    key="pl-PL"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/ru-RU/mock/preview/route"
    hrefLang="ru-RU"
    key="ru-RU"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/th-TH/mock/preview/route"
    hrefLang="th-TH"
    key="th-TH"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/tr-TR/mock/preview/route"
    hrefLang="tr-TR"
    key="tr-TR"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/mock/preview/route"
    hrefLang="x-default"
    key="default"
    rel="alternate"
  />
  <link
    href="https://baseUrl.com/es-ES/mock/preview/route"
    key="canonical"
    rel="canonical"
  />
  <script
    key="ldjson-0"
    type="application/ld+json"
  >
    {"foo":"MOCK_LD_JSON"}
  </script>
  <script
    key="ldjson-1"
    type="application/ld+json"
  >
    {"bar":"MOCK_LD_JSON"}
  </script>
</HelmetWrapper>
`;

exports[`SEOHead robots defaults to robots "on" 1`] = `
<meta
  content="index, follow"
  name="robots"
/>
`;

exports[`SEOHead robots explicity turns robots "on" 1`] = `
<meta
  content="index, follow"
  name="robots"
/>
`;

exports[`SEOHead robots turns robots "off" 1`] = `
<meta
  content="noindex, nofollow"
  name="robots"
/>
`;
